lib/srv/db/mongodb/engine.go

```diff
@@ -0,0 +1,216 @@
+		return trace.Wrap(err)
+	}
+	e.Log.Debugf("<=== %v", serverMessage)
+	// ... and pass it back to the client.
+	_, err = clientConn.Write(serverMessage.GetBytes())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Keep reading if server indicated it has more to send.
+	for serverMessage.MoreToCome(clientMessage) {
+		serverMessage, err = protocol.ReadServerMessage(ctx, serverConn)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		e.Log.Debugf("<=== %v", serverMessage)
+		_, err = clientConn.Write(serverMessage.GetBytes())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// authorizeConnection does authorization check for MongoDB connection about
+// to be established.
+func (e *Engine) authorizeConnection(sessionCtx *common.Session) error {
+	ap, err := e.Auth.GetAuthPreference()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	mfaParams := services.AccessMFAParams{
+		Verified:       sessionCtx.Identity.MFAVerified != "",
+		AlwaysRequired: ap.GetRequireSessionMFA(),
+	}
+	// Only the username is checked upon initial connection. MongoDB sends
+	// database name with each protocol message (for query, update, etc.)
+	// so it is checked when we receive a message from client.
+	err = sessionCtx.Checker.CheckAccessToDatabase(sessionCtx.Server, mfaParams,
+		&services.DatabaseLabelsMatcher{Labels: sessionCtx.Server.GetAllLabels()},
+		&services.DatabaseUserMatcher{User: sessionCtx.DatabaseUser})
+	if err != nil {
+		e.Audit.OnSessionStart(e.Context, sessionCtx, err)
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// authorizeClientMessage checks if the user can run the provided MongoDB command.
+//
+// Each MongoDB command contains information about the database it's run in
+// so we check it against allowed databases in the user's role.
+func (e *Engine) authorizeClientMessage(sessionCtx *common.Session, message protocol.Message) error {
+	// Mongo uses OP_MSG for most operations now.
+	msg, ok := message.(*protocol.MessageOpMsg)
+	if !ok {
+		return nil
+	}
+	// Each message has a database information in it.
+	database := msg.GetDatabase()
+	if database == "" {
+		e.Log.Warnf("No database info in message: %v.", message)
+		return nil
+	}
+	err := sessionCtx.Checker.CheckAccessToDatabase(sessionCtx.Server,
+		services.AccessMFAParams{Verified: true},
+		&services.DatabaseLabelsMatcher{Labels: sessionCtx.Server.GetAllLabels()},
+		&services.DatabaseUserMatcher{User: sessionCtx.DatabaseUser},
+		&services.DatabaseNameMatcher{Name: database})
+	if err != nil {
+		e.Audit.OnQuery(e.Context, sessionCtx, common.Query{
+			Database: msg.GetDatabase(),
+			// Commands may consist of multiple bson documents.
+			Query: strings.Join(msg.GetDocumentsAsStrings(), ", "),
+			Error: err,
+		})
+		return trace.Wrap(err)
+	}
+	e.Audit.OnQuery(e.Context, sessionCtx, common.Query{
+		Database: msg.GetDatabase(),
+		// Commands may consist of multiple bson documents.
+		Query: strings.Join(msg.GetDocumentsAsStrings(), ", "),
+	})
```

Comment:
nit: it can be reduce to single e.Audit.OnQuery call:
```suggestion
	err := sessionCtx.Checker.CheckAccessToDatabase(sessionCtx.Server,
		services.AccessMFAParams{Verified: true},
		&services.DatabaseLabelsMatcher{Labels: sessionCtx.Server.GetAllLabels()},
		&services.DatabaseUserMatcher{User: sessionCtx.DatabaseUser},
		&services.DatabaseNameMatcher{Name: database})
		
	e.Audit.OnQuery(e.Context, sessionCtx, common.Query{
	    Database: msg.GetDatabase(),
	    // Commands may consist of multiple bson documents.
	    Query: strings.Join(msg.GetDocumentsAsStrings(), ", "),
	    Error: err,
	})
	
	if err != nil {
		return trace.Wrap(err)
	}
```